### PR TITLE
Fix: LocalAppTestCase does not emit the correct port in logs (#2668)

### PR DESCRIPTION
**Pull request by baladithyab**
https://github.com/DataBiosphere/azul/issues/2668

### Author

- [x] PR title references issue
- [x] PR title matches issue title (preceded by `Fix: ` for bugs)   <sub>or there is a good reason why they're different</sub>
- [x] Title of main commit references issue
- [x] PR is connected to Zenhub issue and description links to issue

### Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

### Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>
- [x] Added announcement to PR description                  <sub>or this PR does not require announcement</sub>
- [x] Added checklist items for additional operator tasks   <sub>or this PR does not require additional tasks</sub>

### Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR does not touch requirements*.txt</sub>

### Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

### Primary reviewer (after approval)

- [x] Commented in issue about demo expectations            <sub>or labelled issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear              <sub>or issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                               <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [ ] Pushed merge commit to Github

### Operator (after pushing the merge commit)

- [ ] Made announcement requested by author                 <sub>or PR description does not contain an announcement</sub>
- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate       <sub>or or this PR is authored by lead</sub>
- [ ] Pushed merge commit to Gitlab                         <sub>or or merge commit can be pushed later, with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Build passes on Gitlab

### Operator (reindex) 

- [ ] Started reindex in target deployment                  <sub>or this PR does not require reindexing</sub>
- [ ] Checked for and triaged indexing failures             <sub>or this PR does not require reindexing</sub>
- [ ] Emptied fail queues in target deployment              <sub>or this PR does not require reindexing</sub>

### Operator

- [ ] Unassigned PR

### CODE REVIEW

**Review by amarjandu — CHANGES_REQUESTED** [review]
https://github.com/DataBiosphere/azul/runs/4121373735?check_suite_focus=true#step:5:4751

Why does the unittest suite show all these extra `Serving on` statements at the end of the log output?

**Comment by baladithyab** [comment]
> https://github.com/DataBiosphere/azul/runs/4121373735?check_suite_focus=true#step:5:4751
> 
> Why does the unittest suite show all these extra `Serving on` statements at the end of the log output?

@amarjandu I believe this to be due to an instance of LocalAppTestCase being initialized for multiple unit tests.
I will be taking a look on how to reduce this but I do believe that replacing the `print` statement with `log.debug` could help.

File: test/app_test_case.py [hunk 1]
```diff
@@ -38,6 +38,8 @@ def __init__(self, app, config, host, port):
         self.server_wrapper = LocalDevServer(app, config, host, port)
 
     def run(self):
+        log.debug(f'Serving on http://{self.address[0]}:{self.address[1]}')
+        # print(f'Serving on http://{self.address[0]}:{self.address[1]}')
```
Comment (amarjandu) [inline test/app_test_case.py:42]: Why is this print statement still here?

File: test/app_test_case.py [hunk 2]
```diff
@@ -38,6 +38,8 @@ def __init__(self, app, config, host, port):
         self.server_wrapper = LocalDevServer(app, config, host, port)
 
     def run(self):
+        log.debug(f'Serving on http://{self.address[0]}:{self.address[1]}')
```
Comment (amarjandu) [inline test/app_test_case.py:41]: Did you test your changes? the `log` has it's level set to `INFO`. There's no output in my PyCharm instance.

Your changes: 
```
> (cd test && python -m unittest service.test_response)
2021-11-18 11:29:04,006 DEBUG   MainThread: Successfully wrote aggregate for test/files/c18f9014-7086-5216-a73a-e37af51f2bbb with 1 contribution(s).
2021-11-18 11:29:04,973 INFO    Thread-2: Received GET request for '/health/basic', with query null and headers {"host": "127.0.0.1:59904", "user-agent": "python-requests/2.26.0", "accept-encoding": "gzip, deflate, br", "accept": "*/*", "connection": "keep-alive"}.
2021-11-18 11:29:04,973 INFO    Thread-2: Did not authenticate request.
2021-11-18 11:29:04,973 DEBUG   Thread-2: Returning 200 response with headers {"Access-Control-Allow-Origin": "*", "Access-Control-Allow-Headers": "Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key"}. See next line for the first 1024 characters of the body.
{"up": true}
127.0.0.1 - - [18/Nov/2021 11:29:04] "GET /health/basic HTTP/1.1" 200 -
```
INFO
```
> (cd test && python -m unittest service.test_response)
2021-11-18 11:27:11,896 DEBUG   MainThread: Successfully wrote aggregate for test/files/c18f9014-7086-5216-a73a-e37af51f2bbb with 1 contribution(s).
2021-11-18 11:27:12,827 INFO    Thread-1: Serving on http://127.0.0.1:59860
2021-11-18 11:27:12,832 INFO    Thread-2: Received GET request for '/health/basic', with query null and headers {"host": "127.0.0.1:59860", "user-agent": "python-requests/2.26.0", "accept-encoding": "gzip, deflate, br", "accept": "*/*", "connection": "keep-alive"}.
```

**Review by amarjandu — CHANGES_REQUESTED** [review]
@baladithyab the suggested changes don't output anything, I don't see the output in my terminal nor do I see it in the CI test that ran for this commit.

Please test suggested changes locally, if you get stuck on something, give me a ping and we can hop into a session to debug it together.

**Review by amarjandu — APPROVED** [review]
Note: the log messages don't pop up in github actions, but there was consensus that this was OK.

!Squash Rebase plz!

File: test/app_test_case.py [hunk 3]
```diff
@@ -38,6 +38,9 @@ def __init__(self, app, config, host, port):
         self.server_wrapper = LocalDevServer(app, config, host, port)
 
     def run(self):
+        # FIXME: Log should output on line after unit test description
```
Comment (amarjandu) [inline test/app_test_case.py:41]: `FIXME: Log output should be on a newline after unit test description`

**Review by amarjandu — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
The GitHub action logs don't contain the log line that this PR is supposed to question. The [hypothetical cause](https://github.com/DataBiosphere/azul/issues/3665#issuecomment-1017757566) was confirmed to be the fact that the unit test loggers aren't configured properly. Only the logger for the module containing the test is configured (by passing it to `configure_test_logging`) any sibling modules imported by that module won't have their loggers configured (unless they contain a call to `configure_test_logging`). This is the case for some of the children (not descendants) of the `test` directory. They contain support code for other tests and therefore don't call `configure_test_logging` at the top level.

To solve this, we could either

- call `configure_test_logging` in all modules underneath test or
- configure the `test` logger in `configure_test_logging`

We should try the latter first and, if that works, change all calls to `configure_test_logging(log)` to `configure_test_logging()`. That argument is redundant, because the logger referenced by `log` should be a child logger of the `test` logger. What I refer to as *`test` logger* is the logger whose name is `test`.

**Review by amarjandu — APPROVED** [review]
Looks good, confirmed CI test output.

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**